Add task for setting up telemetry gateway (#3625)
istio/istio.io: content/docs/tasks/telemetry/gateways/index.md

---
title: Remotely Accessing Telemetry Add-ons
description: This task shows you how to configure external access to the set of Istio telemetry add-ons.
weight: 99
keywords: [telemetry,gateway,jaeger,zipkin,tracing,kiali,prometheus,add-ons]
---

This task shows how to configure Istio to expose and access the telemetry add-ons outside of
a cluster.

> ## Configuring Remote Access

Reviewer: Don't capitalize headings.

Author: done.

Access to the telemetry add-ons can be configured in a number of different ways. This task will cover
two basic access methods: secure (via https) and insecure (via http). The secure method is *strongly
recommended* for any production or sensitive environment. Insecure access is simpler to set up, but
will not protect any credentials or data transmitted outside of your cluster.

### Option 1: Secure Access (HTTPS)

A server certificate is required for secure access. Follow these steps to install and configure
server certificates for a domain that you control.

> It is possible to use self-signed certificates instead, but that is not covered by this task.

Author: updated.

> [Securing Gateways with HTTPS Using Secret Discovery Service](/docs/tasks/traffic-management/secure-ingress/sds/)

Author: done.

> provides general details on using self-signed certificates to access in-cluster services.

Author: done.

{{< warning >}}

> This task is concerned *solely* with securing the transport layer for access. When exposing the telemetry

Author: updated.

> services externally, the services should be installed and configured to require authentication.

Author: updated.

{{< /warning >}}

> 1. [Install Istio](/docs/setup/) in your cluster with `cert-manager` enabled and `istio-ingressgateway` configured to use

Reviewer 2: Based on the commands you use, I'm assuming you need people to install Istio on a K8s cluster. Please be more specific on the documents you link. I've suggested a change to point to the K8s setup landing page but if there is a specific installation flow you want readers to follow, feel free to change the link accordingly.

```suggestion
1. [Install Istio](/docs/setup/kubernetes/) in your cluster and enable the `cert-manager` flag and configure the `istio-ingressgateway` to use
```

> the [Secret Discovery Service](https://www.envoyproxy.io/docs/envoy/latest/configuration/secret#config-secret-discovery-service).

Reviewer 2: ```suggestion
the [Secret Discovery Service](https://www.envoyproxy.io/docs/envoy/latest/configuration/secret#config-secret-discovery-service).
    To install Istio accordingly, use the following Helm installation options:
```

Author: Done.

>     Use the following helm install options:

Reviewer 2: ```suggestion
```

    * `--set gateways.enabled=true`
    * `--set gateways.istio-ingressgateway.enabled=true`
    * `--set gateways.istio-ingressgateway.sds.enabled=true`
    * `--set certmanager.enabled=true`
    * `--set certmanager.email=mailbox@donotuseexample.com`

>     Install the telemetry add-ons using the following helm install options:

Reviewer: We use 'addons' with no -

Author: Does the [style guide](https://istio.io/about/contribute/style-guide/) need updating?

> Do |	Don’t
 add-on |	addon

Reviewer 2: @douglas-reid You are correct "add-on" is the preferred spelling.

 ```suggestion
    To additionally install the telemetry add-ons, use the following Helm installation options:
```

Author: Ok. updated.

    * Grafana: `--set grafana.enabled=true`
    * Kiali: `--set kiali.enabled=true`
    * Prometheus: `--set prometheus.enabled=true`
    * Tracing: `--set tracing.enabled=true`

1. Configure the DNS records for your domain.

>     1. Get the External IP Address of the `istio-ingressgateway`.

Reviewer 2: ```suggestion
    1. Get the external IP address of the `istio-ingressgateway`.
```

        {{< text bash >}}
        $ kubectl -n istio-system get service istio-ingressgateway -o jsonpath='{.status.loadBalancer.ingress[0].ip}'
        <IP ADDRESS OF CLUSTER INGRESS>
        {{< /text >}}

    1. Set an environment variable to hold your target domain.

        {{< text bash >}}
        $ TELEMETRY_DOMAIN=<your.desired.domain>
        {{< /text >}}

>     1. Point your desired domain at that External IP Address via your domain provider.

Reviewer 2: How should the reader accomplish this? I assume the actual command depends on the provider. Could you provide an example?

Author: This is a topic that goes well beyond the scope of this task and varies from provider to provider. I've added some links to what I think are decently broad example providers FAQs, etc.

    1. Verify that the DNS records are correct.

        {{< text bash >}}
        $ dig +short $TELEMETRY_DOMAIN
        <IP ADDRESS OF CLUSTER INGRESS>
        {{< /text >}}

1. Generate a server certificate

    {{< text bash >}}
    $ cat <<EOF | kubectl apply -f -
    apiVersion: certmanager.k8s.io/v1alpha1
    kind: Certificate
    metadata:
      name: telemetry-gw-cert
      namespace: istio-system
    spec:
      secretName: telemetry-gw-cert
      issuerRef:
        name: letsencrypt
        kind: ClusterIssuer
      commonName: $TELEMETRY_DOMAIN
      dnsNames:
      - $TELEMETRY_DOMAIN
      acme:
        config:
        - http01:
            ingressClass: istio
          domains:
          - $TELEMETRY_DOMAIN
    ---
    EOF
    certificate.certmanager.k8s.io "telemetry-gw-cert" created
    {{< /text >}}

1. Wait until the server certificate is ready.

    {{< text bash >}}
    $ JSONPATH='{range .items[*]}{@.metadata.name}:{range @.status.conditions[*]}{@.type}={@.status}{end}{end}' && kubectl -n istio-system get certificates -o jsonpath="$JSONPATH"
    telemetry-gw-cert:Ready=True
    {{< /text >}}

> 1. Apply networking configuration for the telemetry add-ons.

Reviewer 2: ```suggestion
1. Apply the networking configuration to the telemetry add-ons.
```

>    1. Apply the Grafana configuration.

Reviewer 2: Be specific about to what object/component you are applying the configuration to and with what purpose.

```suggestion
   1. Apply the following configuration to enable Grafana:
```

Author: done.

              mode: SIMPLE
              serverCertificate: sds
              privateKey: sds
              credentialName: telemetry-gw-cert
            hosts:
            - "$TELEMETRY_DOMAIN"
        ---
        apiVersion: networking.istio.io/v1alpha3
        kind: VirtualService
        metadata:
          name: grafana-vs
          namespace: istio-system
        spec:
          hosts:
          - "$TELEMETRY_DOMAIN"
          gateways:
          - grafana-gateway
          http:
          - match:
            - port: 15031
            route:
            - destination:
                host: grafana
                port:
                  number: 3000
        ---
        apiVersion: networking.istio.io/v1alpha3
        kind: DestinationRule
        metadata:
          name: grafana
          namespace: istio-system
        spec:
          host: grafana
          trafficPolicy:
            tls:
              mode: DISABLE
        ---
        EOF
        gateway.networking.istio.io "grafana-gateway" configured
        virtualservice.networking.istio.io "grafana-vs" configured
        destinationrule.networking.istio.io "grafana" configured
        {{< /text >}}

>    1. Apply the Kiali configuration.

Reviewer 2: ```suggestion
   1. Apply the following configuration to enable Kiali.
```

Author: done.

        {{< text bash >}}
        $ cat <<EOF | kubectl apply -f -
        apiVersion: networking.istio.io/v1alpha3
        kind: Gateway
        metadata:

>           name: kilai-gateway

Reviewer 3: typo - "kiali" not "kilai"

Author: fixed.

            tls:
              mode: SIMPLE
              serverCertificate: sds
              privateKey: sds
              credentialName: telemetry-gw-cert
            hosts:
            - "$TELEMETRY_DOMAIN"
        ---
        apiVersion: networking.istio.io/v1alpha3
        kind: VirtualService
        metadata:
          name: kiali-vs
          namespace: istio-system
        spec:
          hosts:
          - "$TELEMETRY_DOMAIN"
          gateways:
          - kiali-gateway
          http:
          - match:
            - port: 15029
            route:
            - destination:
                host: kiali
                port:
                  number: 20001
        ---
        apiVersion: networking.istio.io/v1alpha3
        kind: DestinationRule
        metadata:
          name: kiali
          namespace: istio-system
        spec:
          host: kiali
          trafficPolicy:
            tls:
              mode: DISABLE
        ---
        EOF
        gateway.networking.istio.io "kiali-gateway" configured
        virtualservice.networking.istio.io "kiali-vs" configured
        destinationrule.networking.istio.io "kiali" configured
        {{< /text >}}

>    1. Apply the Prometheus configuration.

Reviewer 2: ```suggestion
   1. Apply the following configuration to enable Prometheus.
```

Author: done.

              serverCertificate: sds
              privateKey: sds
              credentialName: telemetry-gw-cert
            hosts:
            - "$TELEMETRY_DOMAIN"
        ---
        apiVersion: networking.istio.io/v1alpha3
        kind: VirtualService
        metadata:
          name: prometheus-vs
          namespace: istio-system
        spec:
          hosts:
          - "$TELEMETRY_DOMAIN"
          gateways:
          - prometheus-gateway
          http:
          - match:
            - port: 15030
            route:
            - destination:
                host: prometheus
                port:
                  number: 9090
        ---
        apiVersion: networking.istio.io/v1alpha3
        kind: DestinationRule
        metadata:
          name: prometheus
          namespace: istio-system
        spec:
          host: prometheus
          trafficPolicy:
            tls:
              mode: DISABLE
        ---
        EOF
        gateway.networking.istio.io "prometheus-gateway" configured
        virtualservice.networking.istio.io "prometheus-vs" configured
        destinationrule.networking.istio.io "prometheus" configured
        {{< /text >}}

>    1. Apply the tracing configuration.

Reviewer 2: ```suggestion
   1. Apply the following configuration to enable tracing.
```

Author: done.

        ---
        apiVersion: networking.istio.io/v1alpha3
        kind: VirtualService
        metadata:
          name: tracing-vs
          namespace: istio-system
        spec:
          hosts:
          - "$TELEMETRY_DOMAIN"
          gateways:
          - tracing-gateway
          http:
          - match:
            - port: 15032
            route:
            - destination:
                host: tracing
                port:
                  number: 80
        ---
        apiVersion: networking.istio.io/v1alpha3
        kind: DestinationRule
        metadata:
          name: tracing
          namespace: istio-system
        spec:
          host: tracing
          trafficPolicy:
            tls:
              mode: DISABLE
        ---
        EOF
        gateway.networking.istio.io "tracing-gateway" configured
        virtualservice.networking.istio.io "tracing-vs" configured
        destinationrule.networking.istio.io "tracing" configured
        {{< /text >}}

1. Visit the telemetry add-ons via your browser.

   * Kiali: `https://$TELEMERY_DOMAIN:15029/`
   * Prometheus: `https://$TELEMERY_DOMAIN:15030/`
   * Grafana: `https://$TELEMERY_DOMAIN:15031/`

>    * Tracing: `https://$TELEMERY_DOMAIN:15032/`

Reviewer 3: in the above 4 lines, there is a typo in the env var name - its "TELEMETRY" not "TELEMERY"

Author: good catch.

> ### Option 2: Insecure Access (HTTP)

Reviewer 2: Please see my comments on Option 1.
Consider using tabs for the options instead of headings: https://preliminary.istio.io/about/contribute/creating-and-editing-pages/#using-tabs

Author: I tried tabs in an early version of this page for the content below the option level and ran into issues with embedded text blocks (bad renders). Are embedded yaml text blocks supported in tabs?

Author: Yeah, this is the result when I use tabs:

![image](https://user-images.githubusercontent.com/21148125/54163768-4ee3cf80-4417-11e9-8a31-21950e770ae8.png)

Notice all the broken text in what were previously working `{{< text >}}` block.

Reviewer: It's somewhat hit or miss with tabs and nested stuff. We're seeing the same problem with the case of a {{<text>}} block in a bullet inside of a {{<warning>}} or {{<tip>}}. 

The grammars just don't nest properly and the parsers are stepping on one another. The key challenge is that markdown gives meaning to indents and spacing. As data is transformed in the Hugo pipeline, chunks of data suddenly gets treated as a code block (due to indent) leading to bad things downstream.

            - "*"
        ---
        apiVersion: networking.istio.io/v1alpha3
        kind: VirtualService
        metadata:
          name: grafana-vs
          namespace: istio-system
        spec:
          hosts:
          - "*"
          gateways:
          - grafana-gateway
          http:
          - match:
            - port: 15031
            route:
            - destination:
                host: grafana
                port:
                  number: 3000
        ---
        apiVersion: networking.istio.io/v1alpha3
        kind: DestinationRule
        metadata:
          name: grafana
          namespace: istio-system
        spec:
          host: grafana
          trafficPolicy:
            tls:
              mode: DISABLE
        ---
        EOF
        gateway.networking.istio.io "grafana-gateway" configured
        virtualservice.networking.istio.io "grafana-vs" configured
        destinationrule.networking.istio.io "grafana" configured
        {{< /text >}}

   1. Apply the Kiali configuration.

        {{< text bash >}}
        $ cat <<EOF | kubectl apply -f -
        apiVersion: networking.istio.io/v1alpha3
        kind: Gateway
        metadata:

>           name: kilai-gateway

Reviewer 3: typo - "kiali" not "kilai"

Author: thanks. fixed.
